### DISCUSSION TITLE
Method "UpdateSubresources" is not working

### DISCUSSION THREAD

**Discussion by snpkhy** [amerkoleci/Vortice.Windows#448] [Q&A] [answered]
Method "UpdateSubresources" does not get SubresourceData as a parametr, instead of this method wants SubresourceInfo without it i can not create a 2D texture. What to do in this situation... How can I create a 2D Texture and set it on triangle?
`        uint[] texture = GenerateTexture();

        SubresourceData textureData = new SubresourceData()
        {
            Data = (nint)texture[0],
            RowPitch = TextureWidth * TexturePixelSize, 
            SlicePitch = TextureWidth * TexturePixelSize * TextureHeight,
        };
        var subInfo = new SubresourceInfo() 
        {
            RowPitch = (int)textureData.RowPitch,
            DepthPitch = 1,
            Offset = 0
        };
        mCommandList!.UpdateSubresources(mTexture!, textureUploadHeap, 0, 0, 1, &subInfo);
    }

    private uint[] GenerateTexture()
    {
        uint[] pixels =
            [
                0xFFFFFFFF,
                0x00000000,
                0xFFFFFFFF,
                0x00000000,
                0x00000000,
                0xFFFFFFFF,
                0x00000000,
                0xFFFFFFFF,
                0xFFFFFFFF,
                0x00000000,
                0xFFFFFFFF,
                0x00000000,
                0x00000000,
                0xFFFFFFFF,
                0x00000000,
                0xFFFFFFFF,
            ];
        return pixels;
    }`

**Comment by amerkoleci** [comment]
Did you tried to use https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.Direct3D11/ID3D11Device.cs#L811?

**Reply by snpkhy** [reply]
But it is D3D11

**Reply by amerkoleci** [reply]
It is D3D11 indeed, whats wrong?

**Reply by snpkhy** [reply]
I am using D3D12

**Reply by amerkoleci** [reply]
Here you go, complete example for D3D12 texture cube rendering, code here:

https://github.com/amerkoleci/Vortice.Windows.Samples/tree/main/src/Direct3D12/07_DrawTexturedCube

**Reply by snpkhy** [reply]
Cool, but how do i use this in my project to render simple picture :(

**Reply by zaafar** [reply]
> Cool, but how do i use this in my project to render simple picture :(

isn't that something you have to figure out as a programmer working on the project that wants to render simple picture? I just don't understand how the example shared by @amerkoleci  isn't enough?

**Reply by amerkoleci** [reply]
That's what I'm trying to figure out, if he doesn't have experience with graphics, why using something like directx 12?

**Reply by snpkhy** [reply]
Whoa, whoa, whoa, guys... all I wanted was a normal help, not a hate. First of all, if you're such pros, why doesn't a simple project that draws a square and not only that work even in the same tutorial? Secondly, I have experience with graphics, but this is the first time I've worked with DirectX 12, so why let someone down or say that I'm some kind of sucker... I really don't understand why I need to throw an example of texture overlay on a cube, why not on the same triangle or square. It's much easier.... It seems like you don't know how to do it yourself. Based on the code you gave me, I managed to create a square by some miracle, but it seems to me that it's not a square, it's still a cube, it just doesn't rotate and overlay a picture instead of pixels. And the functions were really not working. Thanks to Amer Koleci, who reworked the UpdateSubresources method, it is already possible to create and drop the data into the texture, but for some reason, even if I drop the data there, it is displayed in black. Apparently, most of all, I am not putting the data there correctly.

**Reply by amerkoleci** [reply]
There is no hate, just advices, this are my free time projects and I provide sample projects that I think makes sense not what community asks, so if you need something you can ask or implement yourself. 

Take care!

**Reply by snpkhy** [reply]
Amer Koleci, could you please explain the method I've already asked about, "UpdateSubresources"... Does it put the data into the texture or should I forget about it and try something else? Can you advise me a method that will bring me the desired data and display it on the shape?

**Comment by snpkhy** [comment, marked as answer]
@amerkoleci everything I wanted to do worked. All that helped me was that you fixed the "UpdateSubresorces" method. All the other tutorials are broken :(. If someone else has also encountered this problem, I advise you to use this tutorial:
https://github.com/microsoft/DirectX-Graphics-Samples

**Reply by snpkhy** [reply]
@amerkoleci if u wanna I can make a good tutorial on vortice
